### PR TITLE
Fix CURRENT OF to work with PL/pgSQL cursors.

### PR DESCRIPTION
**Pull request by hlinnaka**
It only worked for cursors declared with DECLARE CURSOR, before. You got
an "there is no parameter $0" error if you tried. This moves the decision
on whether a plan is "simply updatable", from the parser to the planner.
Doing it in the parser was awkward, because we only want to do it for
queries that are used in a cursor, and for SPI queries, we don't know it
at that time yet.

For some reason, the copy, out, read-functions of CurrentOfExpr were missing
the cursor_param field. While we're at it, reorder the code to match
upstream.

This only makes the required changes to the Postgres planner. I'm not sure
what ORCA would need here, although all the existing test cases seem to
pass with ORCA too. Which is actually a bit surprising, since I don't see
how the simplyUpdatable flag can ever get set for ORCA-planned queries now,
nor do I see any code to tell ORCA to add the gp_segment_id and ctid junk
columns to the query. But ah well, if it works, that's good.

Fixes github issue #3292.

### CODE REVIEW

**Comment by hlinnaka** [comment]
We bumped into this while working on the PostgreSQL 8.4 merge. 8.4 added a new cursor-variant of the FOR-statement to PL/pgSQL, and it added a test case very similar to the one added here, that exposed this issue. But as can be seen here, it's reproducable with the OPEN command on master, too.

**Comment by hlinnaka** [comment]
I solved the mystery why updatable cursors seemed to work with ORCA: the ORCA translator doesn't handle DECLARE CURSOR commands at all, and falls back to the Postgres planner. Before this PR, only cursors opened with DECLARE CURSOR were updatable, and it only worked with optimizer=on because of the fallback.

**Comment by pivotal-mike** [comment]
@oarap @vraghavan78 Can we please have a few eyes on this?  It would make the upcoming merge PR a bit nicer.

File: src/backend/executor/execCurrent.c [hunk 1]
```diff
@@ -291,10 +295,8 @@ getCurrentOf(CurrentOfExpr *cexpr,
  * Fetch the string value of a param, verifying it is of type REFCURSOR.
  */
 static char *
-fetch_param_value(ExprContext *econtext, int paramId)
+fetch_param_value(ParamListInfo paramInfo, int paramId)
```
Comment (danielgustafsson) [inline src/backend/executor/execCurrent.c]: Why this change? AFAICT upstream pass an `ExprContext` here and extracts the `ParamListInfo` from it.

**Comment by hlinnaka** [comment]
On 09/21/2017 01:06 PM, Daniel Gustafsson wrote:
> danielgustafsson commented on this pull request.
> 
> :+1: on moving this from parse analysis. LGTM with a few small questions, none of which are blockers.
> 
>> @@ -808,3 +808,35 @@ DECLARE c CURSOR FOR SELECT * FROM aocotest;
>   DELETE FROM aocotest WHERE CURRENT OF c;
>   ERROR:  "aocotest" is not simply updatable
>   ROLLBACK;
> +--
> +-- PL/pgSQL cursors
> +--
> +--  Test that cursors opened in PL/pgSQL can also be updated.
> 
> Perhaps this comment should be expanded with a note saying that it's not supported in ORCA yet? The comment with the below ERROR is currently a bit confusing.
> 
>> @@ -291,10 +295,8 @@ getCurrentOf(CurrentOfExpr *cexpr,
>    * Fetch the string value of a param, verifying it is of type REFCURSOR.
>    */
>   static char *
> -fetch_param_value(ExprContext *econtext, int paramId)
> +fetch_param_value(ParamListInfo paramInfo, int paramId)
> 
> Why this change? AFAICT upstream pass an `ExprContext` here and extracts the `ParamListInfo` from it.

The caller of getCurrentOf(),  in cdbmutate.c, doesn't have an 
ExprContext at hand. Only EState, which contains the ParamListInfo. But 
now that I look at this again, it easily can call 
GetPerTupleExprContext() to get an ExprContext.

Added a commit to do that, to be squashed before pushing. Thanks!

- Heikki

**Comment by danielgustafsson** [comment]
> On 21 Sep 2017, at 13:06, Heikki Linnakangas <notifications@github.com> wrote:
> 
> On 09/21/2017 01:06 PM, Daniel Gustafsson wrote:
> 
> > Why this change? AFAICT upstream pass an `ExprContext` here and extracts the `ParamListInfo` from it.
> 
> The caller of getCurrentOf(), in cdbmutate.c, doesn't have an 
> ExprContext at hand. Only EState, which contains the ParamListInfo. But 
> now that I look at this again, it easily can call 
> GetPerTupleExprContext() to get an ExprContext.
> 
> Added a commit to do that, to be squashed before pushing. Thanks!

Ooh, missed that. That explains the previous coding, and I really like the change with calling GetPerTupleExprContext(). +1 on pushing this version.

**Comment by hlinnaka** [comment]
Ok, pushed, thanks for the review!
